File: ktox-lua-cc/src/main/kotlin/lib/Movement.kt
package lib

import common.turtleBack
import common.turtleDig
import common.turtleDigDown
import common.turtleDigUp
import common.turtleDown
import common.turtleForward
import common.turtleTurnLeft
import common.turtleTurnRight
import common.turtleUp

// Tracks the turtle's position/heading, kept in sync purely by dead
// reckoning (counting moves) — GPS is never required for this to work,
// only optionally used to anchor it to real world coordinates and, when
// available, to sanity-check for drift. Home is a fixed reference: the
// exact block the turtle was standing on when the program started, i.e.
// where the fuel/overflow chests are relative to. See calibrateMovement()
// below: with a working wireless/ender modem + GPS host, home/heading are
// real world coordinates and callers should periodically re-sync via a
// fresh gpsLocate() call as a drift safety check (gated on `gpsEnabled`);
// without one, home is just local (0,0,0) with heading 0 — an arbitrary
// but internally consistent origin, since nothing in this file or
// lib/Chest.kt ever needs home to align with true world coordinates, only
// with itself.
//
// Heading: 0 = north (-z), 1 = east (+x), 2 = south (+z), 3 = west (-x)
// when GPS-anchored; an arbitrary but self-consistent label otherwise.
// `homeHeading` is the direction the turtle originally faced (away from
// the fuel chest) — turning to this heading and then turnAround() faces
// the fuel chest.
enum class MoveSign { FORWARD, BACKWARD }

class Movement(
    startX: Int,
    startY: Int,
    startZ: Int,
    startHeading: Int,
    val homeX: Int,
    val homeY: Int,
    val homeZ: Int,
    val homeHeading: Int,
    val gpsEnabled: Boolean,
) {
    var x: Int = startX
    var y: Int = startY
    var z: Int = startZ
    var heading: Int = startHeading

    fun turnLeft() {
        turtleTurnLeft()
        heading = (heading + 3) % 4
    }

    fun turnRight() {
        turtleTurnRight()
        heading = (heading + 1) % 4
    }

    fun turnAround() {
        this.turnRight()
        this.turnRight()
    }

    // Turns (whichever way is shorter) to face the given absolute heading.
    fun faceHeading(target: Int) {
        val diff = (target - heading + 4) % 4
        if (diff == 3) {
            this.turnLeft()
        } else {
            var turned = 0
            while (turned < diff) {
                this.turnRight()
                turned += 1
            }
        }
    }

    // Digs through an obstruction (once) if the first attempt is blocked.
    // Returns false if still blocked afterward (e.g. bedrock).
    fun forward(): Boolean {
        if (turtleForward()) {
            this.applyDelta(MoveSign.FORWARD)
            return true
        }
        turtleDig()
        if (turtleForward()) {
            this.applyDelta(MoveSign.FORWARD)
            return true
        }
        return false
    }

    fun back(): Boolean {
        if (turtleBack()) {
            this.applyDelta(MoveSign.BACKWARD)
            return true
        }
        return false
    }

    fun up(): Boolean {
        if (turtleUp()) {
            y += 1
            return true
        }
        turtleDigUp()
        if (turtleUp()) {
            y += 1
            return true
        }
        return false
    }

    fun down(): Boolean {
        if (turtleDown()) {
            y -= 1
            return true
        }
        turtleDigDown()
        if (turtleDown()) {
            y -= 1
            return true
        }
        return false
    }

    private fun applyDelta(sign: MoveSign) {
        val delta = if (sign == MoveSign.FORWARD) 1 else -1
        when (heading) {
            0 -> z -= delta
            1 -> x += delta
            2 -> z += delta
            else -> x -= delta
        }
    }

    // Manhattan distance back to home (the turtle's actual starting block,
    // not necessarily (0,0,0)) — used as a fuel safety margin before
    // committing to another move.
    fun distanceHome(): Int {
        val dx = x - homeX
        val dz = z - homeZ
        val dy = y - homeY
        val adx = if (dx < 0) -dx else dx
        val ady = if (dy < 0) -dy else dy
        val adz = if (dz < 0) -dz else dz
        return adx + ady + adz
    }
}

// Attempts one-time GPS calibration: reads world position (this becomes
// "home"), moves forward one block (digging through if blocked) to derive
// heading from the resulting position delta, then reads position again.
// GPS is optional — this never fails the caller. If no modem/GPS host is
// available (or the forced calibration move is blocked, e.g. bedrock on
// every side), it falls back to a local (0,0,0) origin with heading 0 and
// gpsEnabled=false, and the turtle stays exactly where it started (no
// forced move attempted unless the first gpsLocate() already succeeded).
fun calibrateMovement(): Movement {
    val start = gpsLocate()
    if (start == null) {
        return Movement(0, 0, 0, 0, 0, 0, 0, 0, false)
    }
    if (!turtleForward()) {
        turtleDig()
        if (!turtleForward()) {
            return Movement(0, 0, 0, 0, 0, 0, 0, 0, false)
        }
    }
    val after = gpsLocate()
    if (after == null) {
        // Already moved one block by this point with no second reading to
        // derive heading from — treat wherever the turtle now stands as
        // the dead-reckoning origin rather than aborting.
        return Movement(0, 0, 0, 0, 0, 0, 0, 0, false)
    }
    val heading = headingFromDelta(after.x - start.x, after.z - start.z)
    return Movement(after.x, after.y, after.z, heading, start.x, start.y, start.z, heading, true)
}

private fun headingFromDelta(dx: Int, dz: Int): Int {
    return when {
        dz < 0 -> 0
        dx > 0 -> 1
        dz > 0 -> 2
        else -> 3
    }
}

// Heading: 0 = north (-z), 1 = east (+x), 2 = south (+z), 3 = west (-x) —
// see Movement's own doc comment above.
fun headingDx(h: Int): Int {
    return if (h == 1) {
        1
    } else if (h == 3) {
        -1
    } else {
        0
    }
}

fun headingDz(h: Int): Int {
    return if (h == 2) {
        1
    } else if (h == 0) {
        -1
    } else {
        0
    }
}

// Axis-aligned navigation to an absolute world position: Y first, then X,
// then Z. Digs through anything in the way (via Movement's forward/up/down)
// — but forward()/up()/down() themselves return false when genuinely
// blocked (e.g. bedrock: dig succeeds on a pocket between two bedrock
// blocks, but the move into the next one still fails). Every step here
// checks that and bails out immediately instead of re-attempting the same
// blocked move in an infinite loop with the target never getting closer.
//
// Returns true if the target was fully reached, false if it stopped
// early — callers that care why (bedrock vs. something else) can inspect
// m.x/m.y/m.z afterward; callers that don't care can ignore the result,
// same as before this returned Unit.
fun navigateTo(m: Movement, targetX: Int, targetY: Int, targetZ: Int): Boolean {
    while (m.y < targetY) {
        if (!m.up()) {
            return false
        }
    }
    while (m.y > targetY) {
        if (!m.down()) {
            return false
        }
    }
    if (m.x != targetX) {
        val toward = if (targetX > m.x) 1 else 3
        m.faceHeading(toward)
        while (m.x != targetX) {
            if (!m.forward()) {
                return false
            }
        }
    }
    if (m.z != targetZ) {
        val toward = if (targetZ > m.z) 2 else 0
        m.faceHeading(toward)
        while (m.z != targetZ) {
            if (!m.forward()) {
                return false
            }
        }
    }
    return true
}

File: src/pkg/player/8durt/lib/Movement.lua
-- package: lib

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "lib/Movement.kt", {["1-41"]=1,["42"]=59,["43-46"]=60,["47"]=64,["48-51"]=65,["52"]=69,["53-57"]=70,["58"]=75,["59"]=76,["60-61"]=77,["62"]=79,["63"]=80,["64"]=81,["65-71"]=82,["72"]=90,["73"]=91,["74-75"]=92,["76"]=94,["77-78"]=95,["79"]=97,["80"]=98,["81-82"]=99,["83-87"]=101,["88"]=105,["89"]=106,["90-91"]=107,["92-96"]=109,["97"]=113,["98"]=114,["99-100"]=115,["101"]=117,["102-103"]=118,["104"]=120,["105"]=121,["106-107"]=122,["108-112"]=124,["113"]=128,["114"]=129,["115-116"]=130,["117"]=132,["118-119"]=133,["120"]=135,["121"]=136,["122-123"]=137,["124-128"]=139,["129"]=143,["130"]=144,["131-132"]=145,["133-134"]=146,["135-136"]=147,["137-142"]=148,["143"]=156,["144"]=157,["145"]=158,["146"]=159,["147"]=160,["148"]=161,["149-158"]=162,["159"]=175,["160"]=176,["161-162"]=177,["163"]=179,["164"]=180,["165"]=181,["166-168"]=182,["169"]=185,["170"]=186,["171-172"]=190,["173"]=192,["174-180"]=193,["181-196"]=197,["197-202"]=208,["203-211"]=218,["212"]=240,["213"]=241,["214-216"]=242,["217"]=245,["218"]=246,["219-221"]=247,["222"]=250,["223"]=251,["224"]=252,["225"]=253,["226"]=254,["227-230"]=255,["231"]=259,["232"]=260,["233"]=261,["234"]=262,["235"]=263,["236-239"]=264,["240-242"]=268}, "lib")

---@alias MoveSign "FORWARD" | "BACKWARD"
local MoveSign = {}

MoveSign.FORWARD = "FORWARD"
MoveSign.BACKWARD = "BACKWARD"

---@class Movement
---@field homeX number
---@field homeY number
---@field homeZ number
---@field homeHeading number
---@field gpsEnabled boolean
---@field x number
---@field y number
---@field z number
---@field heading number
---@field blocksDug number
Movement = {}
Movement.__index = Movement

function Movement:new(startX, startY, startZ, startHeading, homeX, homeY, homeZ, homeHeading, gpsEnabled)
    local self = setmetatable({}, Movement)
    self.homeX = homeX
    self.homeY = homeY
    self.homeZ = homeZ
    self.homeHeading = homeHeading
    self.gpsEnabled = gpsEnabled
    self.x = startX
    self.y = startY
    self.z = startZ
    self.heading = startHeading
    self.blocksDug = 0
    return self
end

function Movement:turnLeft()
    turtle.turnLeft()
    self.heading = (self.heading + 3) % 4
end

function Movement:turnRight()
    turtle.turnRight()
    self.heading = (self.heading + 1) % 4
end

function Movement:turnAround()
    self:turnRight()
    self:turnRight()
end

---@param target number
function Movement:faceHeading(target)
    local diff = (target - self.heading + 4) % 4
    if diff == 3 then
        self:turnLeft()
    else
        local turned = 0
        while turned < diff do
            self:turnRight()
            turned = ktox_plusAssign(turned, 1)
        end
    end
end

---@return boolean
function Movement:forward()
    if turtle.forward() then
        self:applyDelta(MoveSign.FORWARD)
        return true
    end
    if turtle.dig() then
        self.blocksDug = ktox_plusAssign(self.blocksDug, 1)
    end
    if turtle.forward() then
        self:applyDelta(MoveSign.FORWARD)
        return true
    end
    return false
end

---@return boolean
function Movement:back()
    if turtle.back() then
        self:applyDelta(MoveSign.BACKWARD)
        return true
    end
    return false
end

---@return boolean
function Movement:up()
    if turtle.up() then
        self.y = ktox_plusAssign(self.y, 1)
        return true
    end
    if turtle.digUp() then
        self.blocksDug = ktox_plusAssign(self.blocksDug, 1)
    end
    if turtle.up() then
        self.y = ktox_plusAssign(self.y, 1)
        return true
    end
    return false
end

---@return boolean
function Movement:down()
    if turtle.down() then
        self.y = ktox_minusAssign(self.y, 1)
        return true
    end
    if turtle.digDown() then
        self.blocksDug = ktox_plusAssign(self.blocksDug, 1)
    end
    if turtle.down() then
        self.y = ktox_minusAssign(self.y, 1)
        return true
    end
    return false
end

---@param sign MoveSign
function Movement:applyDelta(sign)
    local delta = (sign == MoveSign.FORWARD and 1 or -1)
    if self.heading == 0 then
        self.z = ktox_minusAssign(self.z, delta)
    elseif self.heading == 1 then
        self.x = ktox_plusAssign(self.x, delta)
    elseif self.heading == 2 then
        self.z = ktox_plusAssign(self.z, delta)
    else
        self.x = ktox_minusAssign(self.x, delta)
    end
end

---@return number
function Movement:distanceHome()
    local dx = self.x - self.homeX
    local dz = self.z - self.homeZ
    local dy = self.y - self.homeY
    local adx = (dx < 0 and -dx or dx)
    local ady = (dy < 0 and -dy or dy)
    local adz = (dz < 0 and -dz or dz)
    return adx + ady + adz
end

function Movement:toString()
    return "lib.Movement"
end
Movement.__tostring = function(a) return a:toString() end

---@return Movement
function calibrateMovement()
    local start = gpsLocate()
    if start == nil then
        return Movement:new(0, 0, 0, 0, 0, 0, 0, 0, false)
    end
    if not turtle.forward() then
        turtle.dig()
        if not turtle.forward() then
            return Movement:new(0, 0, 0, 0, 0, 0, 0, 0, false)
        end
    end
    local after = gpsLocate()
    if after == nil then
        return Movement:new(0, 0, 0, 0, 0, 0, 0, 0, false)
    end
    local heading = headingFromDelta(after.x - start.x, after.z - start.z)
    return Movement:new(after.x, after.y, after.z, heading, start.x, start.y, start.z, heading, true)
end

---@param dx number
---@param dz number
---@return number
function headingFromDelta(dx, dz)
    return (function()
        if dz < 0 then
            return 0
        elseif dx > 0 then
            return 1
        elseif dz > 0 then
            return 2
        else
            return 3
        end
    end)()
end

---@param h number
---@return number
function headingDx(h)
    return (h == 1 and 1 or (h == 3 and -1 or 0))
end

---@param h number
---@return number
function headingDz(h)
    return (h == 2 and 1 or (h == 0 and -1 or 0))
end

---@param m Movement
---@param targetX number
---@param targetY number
---@param targetZ number
---@return boolean
function navigateTo(m, targetX, targetY, targetZ)
    while m.y < targetY do
        if not m:up() then
            return false
        end
    end
    while m.y > targetY do
        if not m:down() then
            return false
        end
    end
    if m.x ~= targetX then
        local toward = (targetX > m.x and 1 or 3)
        m:faceHeading(toward)
        while m.x ~= targetX do
            if not m:forward() then
                return false
            end
        end
    end
    if m.z ~= targetZ then
        local toward = (targetZ > m.z and 2 or 0)
        m:faceHeading(toward)
        while m.z ~= targetZ do
            if not m:forward() then
                return false
            end
        end
    end
    return true
end

